docs: tw-1011-batchposter (#3535)
OffchainLabs/arbitrum-docs: docs/how-arbitrum-works/deep-dives/batchposter.mdx

---
title: The batch poster
description: 'Learn the fundamentals of the Arbitrum batch poster.'
author: pete-vielhaber
sme: Mehdi
user_story: As a current or prospective Arbitrum user, I need to learn more about the batch poster.
content_type: reference
---

> Batch posting is a fundamental process for the <a data-quicklook-from="sequencer">Sequencer's</a> operation in Arbitrum. It involves collecting multiple <a data-quicklook-from="child-chain">child chain</a> <a data-quicklook-from="transaction">transactions</a>, organizing them into <a data-quicklook-from="batch">batches</a>, compressing the data to reduce size, and sending these batches to the <a data-quicklook-from="sequencer-inbox">Sequencer Inbox</a> contract on the <a data-quicklook-from="parent-chain">parent chain</a>. This mechanism is crucial for ensuring that transactions are securely recorded on the parent chain while optimizing for costs and performance.

Reviewer: ```suggestion
Batch posting is a fundamental process for the <a data-quicklook-from="sequencer">Sequencer's</a> operation in Arbitrum. It involves collecting multiple <a data-quicklook-from="child-chain">child chain</a> <a data-quicklook-from="transaction">transactions</a>, organizing them into <a data-quicklook-from="batch">batches</a>, compressing the data to reduce size, and sending these batches to the <a data-quicklook-from="sequencer-inbox">Sequencer Inbox</a> contract on the <a data-quicklook-from="parent-chain">parent chain</a>. This mechanism securely records transactions on the parent chain while optimizing for costs and performance.
```

#### 2. The poster decides where it stands

The main loop (`MaybePostSequencerBatch`) first does a safety check (refuse to post if a prior batch reverted), then asks “what batch number and <a data-quicklook-from="L1">L1</a> nonce am I on?” and establishes the L1 time/block window the batch will use. That window comes from reading the L1 SequencerInbox contract’s `MaxTimeVariation`—a read-only handoff out to L1.

#### 3. Messages become a compressed batch

The poster receives messages in order, turning each into typed segments (the transaction itself, plus small bookkeeping notes for <a data-quicklook-from="delayed-inbox">delayed-message</a> markers and time/block advances) and streaming them through Brotli. It watches size and segment-count caps, and adapts compression effort to how far behind it is—more compression when calm, less when backlogged. When it decides the batch is ready (full enough, old enough, or a delayed message is about to expire), it seals and recompresses it. This is the core work done inside the batch poster.

#### 4. The data is routed to storage (DA handoff)

> The sealed bytes are handed to a data availability (DA) provider—an offchain <a data-quicklook-from="data-availability-committee-dac">AnyTrust</a>/<a data-quicklook-from="data-availability-server-das">DAS</a> committee or a custom backend—via the small Writer.Store interface, with a fallback to putting data directly on Ethereum (calldata or <a data-quicklook-from="blob">blobs</a>). What comes back is either a <a data-quicklook-from="data-availability-certificate">certificate</a> (offchain) or the raw data (EthDA). The committee aggregation, <a data-quicklook-from="bls-signature">BLS signing</a>, and blob KZG cryptography all happen behind that interface—handoffs that the batch poster isn’t involved in.

Reviewer: ```suggestion
The sealed bytes are handed to a data availability (DA) provider—an offchain <a data-quicklook-from="data-availability-committee-dac">AnyTrust</a>/<a data-quicklook-from="data-availability-server-das">Data Availability Server (DAS)</a> committee or a custom backend—via the small Writer.Store interface, with a fallback to putting data directly on Ethereum (calldata or <a data-quicklook-from="blob">blobs</a>). What comes back is either a <a data-quicklook-from="data-availability-certificate">certificate</a> (offchain) or the raw data (EthDA). The committee aggregation, <a data-quicklook-from="bls-signature">BLS signing</a>, and blob KZG cryptography all happen behind that interface—handoffs that the batch poster isn’t involved in.
```

Author: dup

#### 6. The transaction is shipped to L1 (`DataPoster` handoff)

The batch poster hands the calldata, blobs, and gas estimate to the `DataPoster`—the “shipping department.” From here, nonce selection, fee bidding, signing, queue persistence, and replace-by-fee-if-stuck are the `DataPoster`’s job, deliberately kept separate so the batch poster never has to deal with Ethereum gas mechanics. The `DataPoster` ultimately submits the transaction to the L1 `SequencerInbox` contract.

#### 7. Finality and feedback

Once that L1 transaction confirms, the transactions cross from the Sequencer’s soft finality into [hard finality](/how-arbitrum-works/reference/finality-and-reorgs.mdx) anchored on Ethereum (L1). The poster updates its backlog estimate (feeding back into compression effort and fee urgency), and a background revert watcher monitors L1—if a posted batch ever fails onchain, it halts all posting for operator intervention.

## Related resources

> - [The Sequencer and censorship resistance](/how-arbitrum-works/deep-dives/sequencer.mdx#batch-posting): batching, compression, and the Sequencer Inbox methods.

Reviewer: ```suggestion
- [The Sequencer and censorship resistance](/how-arbitrum-works/deep-dives/sequencer.mdx#batch-posting): batching, compression, and the <a data-quicklook-from="sequencer-inbox">Sequencer Inbox</a> methods.
```

Author: dup
